src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts

```diff
@@ -143,6 +144,11 @@ export class TerminalInstance implements ITerminalInstance {
 		if (platform.isWindows) {
 			this._processReady.then(() => {
 				this._windowsShellHelper = new WindowsShellHelper(this._processId, this._shellLaunchConfig.executable);
+				this._onCheckWindowsShell = new Emitter<TPromise<string>>();
+				debounceEvent(this._onCheckWindowsShell.event, (l, e) => e, 100, true)(() => {
+					this.checkWindowShell();
+				});
+				this.onData(() => this._onCheckWindowsShell.fire());
```

Comment:
You're right I got a little confused there. But doesn't that mean it will not fire at the beginning if output is returned within 100ms because onEnter is being debounced on the same event? We might actually want to throttle onEnter (fire immediate, but no more than every 100ms), and debounce onData (fire when output stops for 100ms).

It also occurs to me that onData will fire when you press any key while typing in the prompt which probably isn't what we want. Imagine typing "a" (wait 150ms) "b" (wait 150ms) "c": that would fire 3 processes because of the onData handler.

A more efficient approach could be to hook into when a new line is started in the terminal, this would require some new API in xterm.js though. You could probably test out the approach by monkey patching xterm.js temporarily to see if it does what we want. Something like this:

```ts
const originalLineFeed = (<any>this._xterm).inputHandler.lineFeed;
(<any>this._xterm).inputHandler.lineFeed = function () {
  // ...fire event...
  originalLineFeed();
}